lib/modules/home/components/body.dart

```diff
@@ -6,44 +6,63 @@ class Body extends StatelessWidget {
   Widget build(BuildContext context) {
     final state = context.findAncestorStateOfType<_HomeViewImplState>()!;
-    return Screen(
-      body: Container(
-        color: Colors.black,
-        child: StreamsSelector0<bool>.value(
-          stream: state._isLoading,
-          builder: (_, isLoading, child) => Skeleton(
-            isShown: isLoading,
-            baseColor: Colors.white.withOpacity(0.12),
-            highlightColor: Colors.white.withOpacity(0.5),
-            borderRadius: BorderRadius.circular(16),
-            height: 18,
-            child: child!,
-          ),
-          child: Stack(
-            children: [
-              StreamsSelector0<List<SurveyInfo>>.value(
-                stream: state._surveys,
-                builder: (_, surveys, __) => CarouselSlider.builder(
-                  carouselController: state._pageController,
-                  itemCount: surveys.length,
-                  itemBuilder: (_, index, ___) => SlideItem(
-                    survey: surveys[index],
+    return StreamsSelector0<bool>.value(
+      stream: state.isProgressHUDShown,
+      builder: (_, isShown, child) => ProgressHUD(
+        isShow: isShown,
+        child: child!,
+      ),
+      child: Screen(
+        body: SimpleGestureDetector(
+          onVerticalSwipe: (direction) {
+            if (direction != SwipeDirection.down) return;
+            state.delegate?.didSwipeDown.add(null);
+          },
+          child: Container(
+            color: Colors.black,
+            child: StreamsSelector0<bool>.value(
+              stream: state._isLoading,
+              builder: (_, isLoading, child) => Skeleton(
+                key: HomeView.skeletonKey,
+                isShown: isLoading,
+                baseColor: Colors.white.withOpacity(0.12),
+                highlightColor: Colors.white.withOpacity(0.5),
+                child: child!,
+              ),
+              child: Stack(
+                fit: StackFit.expand,
+                children: [
+                  StreamsSelector0<List<SurveyInfo>>.value(
+                    stream: state._surveys,
+                    builder: (_, surveys, __) => surveys.isNotEmpty
+                        ? CarouselSlider.builder(
+                            carouselController: state._pageController,
+                            itemCount: surveys.length,
+                            itemBuilder: (_, index, ___) => SlideItem(
+                              survey: surveys[index],
+                            ),
+                            options: CarouselOptions(
+                              viewportFraction: 1,
+                              initialPage: state._currentPage.value,
+                              height: double.infinity,
+                              onPageChanged: (index, _) =>
+                                  state._currentPage.add(index),
+                            ),
+                          )
+                        : SlideItem.empty(),
```

Comment:
I think we should revise the logic here, if we do not get any surveys, or an error happens, the loading will show forever. We should display error text or some thing like that to let user know it's failed to fetch survey 🤔

Reply:
@manh-t  The reason I put an empty slide here to create a placeholder for skeleton animation.
For alert an error when failed to fetch, I handle it [here](https://github.com/markgravity/flutter-survey/blob/32f0fd130d7f5885a3ac68786d75fe78c81377a8/lib/modules/home/home_presenter.dart#L95)

Reply:
@markgravity ah I see, thanks for letting me know 🙏